### PR TITLE
invoices: reduce heap usage when starting the expiry watcher

### PR DESCRIPTION
**Pull request by bhandras**
This PR reduces heap usage upon start by only caching `invoiceExpiry` objects rather than all active invoices before pushing them to the watcher. Although this is a one time thing, it may improve start time for nodes with large amount of invoices.

### CODE REVIEW

File: invoices/invoice_expiry_watcher.go [hunk 1]
```diff
@@ -128,48 +129,23 @@ func (ew *InvoiceExpiryWatcher) prepareInvoice(
...
-		newInvoiceExpiry := ew.prepareInvoice(paymentHash, invoice)
-		if newInvoiceExpiry != nil {
-			invoicesWithExpiry = append(
-				invoicesWithExpiry, newInvoiceExpiry,
-			)
-		}
-	}
-
-	if len(invoicesWithExpiry) > 0 {
+func (ew *InvoiceExpiryWatcher) AddInvoices(invoices []*invoiceExpiry) {
```
Comment (cfromknecht) [inline invoices/invoice_expiry_watcher.go:132]: nit: should the godoc explain that all entries must be non-nil?

File: invoices/invoice_expiry_watcher.go [hunk 2]
```diff
@@ -128,48 +129,23 @@ func (ew *InvoiceExpiryWatcher) prepareInvoice(
...
-		log.Debugf("Adding invoice '%v' to expiry watcher,"+
-			"expiration: %v", paymentHash, newInvoiceExpiry.Expiry)
-
-		select {
-		case ew.newInvoices <- []*invoiceExpiry{newInvoiceExpiry}:
-		// Select on quit too so that callers won't get blocked in case
-		// of concurrent shutdown.
-		case <-ew.quit:
-		}
+// AddInvoice adds a new invoice to the InvoiceExpiryWatcher.
```
Comment (cfromknecht) [inline invoices/invoice_expiry_watcher.go:145]: nit: godoc here could explain that nil input is a nop

File: invoices/invoice_expiry_watcher.go [hunk 3]
```diff
@@ -103,10 +103,11 @@ func (ew *InvoiceExpiryWatcher) Stop() {
 	}
 }
 
-// prepareInvoice checks if the passed invoice may be canceled and calculates
-// the expiry time.
-func (ew *InvoiceExpiryWatcher) prepareInvoice(
```
Comment (cfromknecht) [inline invoices/invoice_expiry_watcher.go:108]: nice removal of unnecessary member function 👍

**Review by cfromknecht — APPROVED** [review]
@bhandras LGTM, just some minor comments on docs :)

File: invoices/invoice_expiry_watcher.go (hunk 1 above)
Comment (bhandras) [reply invoices/invoice_expiry_watcher.go:132]: Decided to instead make sure that even if we push `nil` entries, they won't be used (see select at the end)

File: invoices/invoice_expiry_watcher.go (hunk 2 above)
Comment (bhandras) [reply invoices/invoice_expiry_watcher.go:146]: So now, even if we were to add a `nil` entry, it'll be discarded in the `select` (below).

File: invoices/invoice_expiry_watcher.go [hunk 4]
```diff
@@ -128,49 +129,23 @@ func (ew *InvoiceExpiryWatcher) prepareInvoice(
...
-		select {
-		case ew.newInvoices <- []*invoiceExpiry{newInvoiceExpiry}:
-		// Select on quit too so that callers won't get blocked in case
-		// of concurrent shutdown.
-		case <-ew.quit:
-		}
-	}
+// AddInvoice adds a new invoice to the InvoiceExpiryWatcher.
+func (ew *InvoiceExpiryWatcher) AddInvoice(invoice *invoiceExpiry) {
+	ew.AddInvoices([]*invoiceExpiry{invoice})
```
Comment (halseth) [inline invoices/invoice_expiry_watcher.go:148]: could also make invoices a variadic parameter and call `AddInvoices(invoice)` directly

**Review by halseth — APPROVED** [review]
LGTM ✅

File: invoices/invoice_expiry_watcher.go (hunk 4 above)
Comment (bhandras) [reply invoices/invoice_expiry_watcher.go]: Good idea! Done.

**Pull request merged**